### PR TITLE
(DOCSP-30333 & DOCSP- 30344): Update Kotlin Read page 

### PR DESCRIPTION
**Pull request by cbullinger**
## Pull Request Info

Jira ticket: 
https://jira.mongodb.org/browse/DOCSP-30333
https://jira.mongodb.org/browse/DOCSP-30344

- [Read Realm Objects - Kotlin SDK](https://preview-mongodbcbullinger.gatsbyjs.io/realm/docsp-30333-filter-data/sdk/kotlin/realm-database/crud/read/): 
   - Add missing sections and corresponding code examples
   - Update existing code examples to use "global" object types from `Schema.kt`
   - Delete unused code examples
   - Add information on filtering and include links to RQL page where appropriate (instead of creating a Filter Data page for Kotlin)

### Reminder Checklist

Before merging your PR, make sure to check a few things.

- [x] Did you tag pages appropriately?
  - genre
  - meta.keywords
  - meta.description
- [ ] Describe your PR's changes in the Release Notes section
- [x] Create a Jira ticket for related docs-app-services work, if any

### Release Notes

- **Kotlin** SDK
  - Realm/Read & Write Data/Read:     

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 1]
```diff
@@ -1,201 +1,654 @@
...
+
+   .. literalinclude:: /examples/generated/kotlin/ReadTest.snippet.synchronous-query.kt
+      :language: kotlin
+
+Find Database Objects
+---------------------
+
+To find objects stored within a database:
+
+1. Pass the object type as a type parameter to
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:141]: I think if you want to present this as a list, it should be an unordered list. I initially viewed it as a set of steps, but some of these are optional parameters you can use with `.query()`. Or maybe it should be something like:

> To find objects...
> 1. Pass the object type... optionally:
>   - Chain queries if you need to further filter the results...
>   - Specify sort order...
> 2. Execute the query...
>
> Work with the query...

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 2]
```diff
@@ -1,201 +1,654 @@
...
+#. Execute the query using either: 
+
+   -  `find() <{+kotlin-local-prefix+}io.realm.kotlin.query/find.html>`__ 
+      for synchronous queries. Returns a collection of results.  
+   -  `asFlow() <{+kotlin-local-prefix+}io.realm.kotlin.query/-realm-element-query/as-flow.html>`__ f
+      or asynchronous queries. Subscribes to a ``Flow`` of results changes. 
+
+   .. tip:: Avoid ``find()`` on the UI Thread
+      
+      ``find()`` runs a synchronous query on the thread it is called from.
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:164]: This is probably nitpicky, but do we really want to blanket recommend against ``find()`` on a UI thread? I assume there may be cases where devs want to perform a query synchronously on the UI thread, and if the data set is small, the impact should be negligible. It's probably good to note the synchronous vs. async usage relative to the UI, but not necessarily make a blanket recommendation to avoid ``find()``. Maybe we would want to say something like "Prefer ``asFlow()`` when working with large data sets on the UI thread to avoid negative UI impact."

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 3]
```diff
@@ -1,201 +1,654 @@
...
 
-.. _kotlin-find-object-by-primary-key:
+Because of the SDK's frozen architecture, you aren't always working 
+with the latest version of an object or collection. 
 
-Find Object by Primary Key
-~~~~~~~~~~~~~~~~~~~~~~~~~~
+To find the latest version of an object or collection, you can call 
+`findLatest() <{+kotlin-local-prefix+}io.realm.kotlin/-mutable-realm/find-latest.html>`__ 
+from a ``MutableRealm`` instance. This method returns a live copy of the
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:183]: If you take my suggestion above about avoiding "returns" if we're talking about working with live data within a write transaction, I'd reword this section a little bit.

**Review by dacharyc — COMMENTED** [review]
I've only gotten to "Query All Objects of a Type," but I wanted to get you these initial comments for feedback.

I am really happy with the way this page is shaping up! This is a really nice, comprehensive resource - thank you for taking the time on this!

File: examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/ReadTest.kt [hunk 4]
```diff
@@ -2,187 +2,892 @@ package com.mongodb.realm.realmkmmapp
...
+                val frogsFavoriteThing = findFrog.favoriteThing // Int
+
+                // Using the correct getter method returns the value
+                val frogsFavoriteNumber = frogsFavoriteThing?.asInt()
+                println("${findFrog.name} likes the number $frogsFavoriteNumber")
+                assertEquals(42, frogsFavoriteNumber) // :remove:
+
+                val thrownException = assertFailsWith<Exception> { // :remove:
+                // Using the wrong getter method throws an exception
+                val frogsFavoriteUUID = frogsFavoriteThing?.asRealmUUID()
```
Comment (dacharyc) [inline examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/ReadTest.kt:332]: Just flagging this as another bit of code that we might want to omit if we're sticking with "don't show code that throws exceptions in code examples."

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 5]
```diff
@@ -1,201 +1,654 @@
...
-have :ref:`Full-Text Search Indexes <kotlin-fts-index>` (FTS) on them. FTS supports Boolean matches and not relevance-based matches.
-To query these properties, use the ``TEXT`` predicate.
+Unlike other properties, you must extract a ``RealmAny`` property's stored 
+value. You can return the currently stored type using 
+`RealmAny.type() <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-any/type.html>`__, 
+then extract the value using the getter method for that type. If you use the
+wrong getter method, the SDK throws an exception. For a full list of getter 
+methods, refer to the 
+`RealmAny <{+kotlin-local-prefix+}io.realm.kotlin.types/-realm-any/index.html>`__ 
+API reference.
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:333]: I might recommend throwing a line here about using conditional expressions to handle multiple types. I like that we show the example of this as a "Tip" because it's not truly core to the functionality - it is, actually, a tip. But suggesting it in a line in this paragraph would clearly state this recommendation and put it in a developer's head.

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 6]
```diff
@@ -1,201 +1,654 @@
...
+
+- ``RealmList``
+- ``RealmSet``
+- ``RealmDictionary`` 
+
+You can query these collection properties the same way you would any other 
+property using RQL. You can also use Kotlin's built-in collection functions
+to filter, sort, and iterate over the results.
+
+RealmList
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:431]: I wonder if we want to increase the OTP page depth to get these headers in the OTP? I know we refer to Lists, Sets, and Map as collection types, but not showing the explicit headers in the OTP adds a layer of cognitive load for folks who are just scanning for that specific type.

File: source/examples/generated/kotlin/ReadTest.snippet.query-remapped-property.kt [hunk 7]
```diff
@@ -0,0 +1,5 @@
+val queryKotlinName = realm.query<Frog>("species == $0", "Muppetarium Amphibius").find().first()
+val queryRemappedName = realm.query<Frog>("latin_name == $0", "Muppetarium Amphibius").find().first()
```
Comment (dacharyc) [inline source/examples/generated/kotlin/ReadTest.snippet.query-remapped-property.kt:2]: Nice! I think explicitly showing this makes this a little clearer.

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 8]
```diff
@@ -1,201 +1,654 @@
...
+   
+   .. literalinclude:: /examples/generated/kotlin/ReadTest.snippet.persisted-name-inverse.kt
+      :language: kotlin
+      :emphasize-lines: 1
+   
+   .. literalinclude:: /examples/generated/kotlin/ReadTest.snippet.query-inverse-persisted-name.kt
+      :language: kotlin
+      :emphasize-lines: 3
+
+Query Geospatial Data
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:542]: Nice! Good call to include this section here.

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 9]
```diff
@@ -1,201 +1,654 @@
...
+Kotlin SDK version 1.11.0 and later adds experimental geospatial APIs that 
+support querying with geospatial data. For more information on using 
+geospatial data in your app, refer to :ref:`<kotlin-geospatial>`.
+
+To query against geospatial data, use the ``GEOWITHIN`` operator. This method 
+takes the ``coordinates`` property of an embedded object and checks if that 
+point is contained within the shape of a given geodata region. For example, 
+you can find all coffee shops (objects with embedded coordinates) within 15 km 
+of a location (the defined geodata region). The format for querying geospatial 
+data is the same, regardless of the shape of the geodata region.
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:556]: I wonder if it's worth mentioning somewhere in here that the SDK provides several geospatial data types to simplify querying with different region and point parameters. Our generic callout above is just about "using geospatial data in your app" - but since this is the "Read" page, I think an explicit mention of the geo data region and points for queries gives users more context about what's possible, which more strongly incentivizes looking at the related page.

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 10]
```diff
@@ -1,201 +1,654 @@
...
+
+.. important:: 
+   
+   You *cannot* use parameterized queries with geospatial data.
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:560]: Nit: this is the only place on the page where we refer to "parameterized queries." If you want to include this note here, we should probably explain parameterized queries up above (maybe with a link out to [the RQL page section](https://www.mongodb.com/docs/realm/realm-query-language/#parameterized-queries) or even just make that section an include that you could also use here) _or_ I think we should omit this from this page.

But this also reminds me that we don't have a geospatial section on the RQL page yet - I wonder if you might want to grab that ticket from Lindsey as I think she has other stuff on her plate: https://jira.mongodb.org/browse/DOCSP-34561

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 11]
```diff
@@ -1,201 +1,654 @@
...
+within the specified ``Geocircle`` region:
+
+.. literalinclude:: /examples/generated/kotlin/Geospatial.snippet.geocircle-query.kt
+   :language: kotlin
+
 .. _kotlin-sort-queries:
 
-Sort Results
-------------
+Sort and Limit Results
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:573]: This is great - really nice work here exploring the options and showing how these work individually and together!

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 12]
```diff
@@ -1,201 +1,654 @@
...
+In the following example, we sort and limit using convenience methods only, 
+RQL only, and then a combination of both to return the same results:
+ 
+.. io-code-block::
+
+   .. input:: /examples/generated/kotlin/ReadTest.snippet.sort-results.kt
+      :language: kotlin
+      :emphasize-lines: 8-10, 16, 23-25
+
+   .. output::
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:601]: Nice work! I think this is a great use of io-code-block with output to clearly illustrate the example.

**Review by dacharyc — COMMENTED** [review]
Just a handful more comments. Really great work! I'll take another look once you've applied any feedback you want to incorporate just to give it one final pass.

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 13]
```diff
@@ -1,201 +1,654 @@
...
-````````````````````````````````````
+.. tip:: Device Sync Always Uses _id as Primary Key
+   
+   If you use Atlas Device Sync, you can always query by the primary key 
+   field ``_id``. This is because the Device Sync data model requires 
+   objects have a primary key named ``_id``. For more information, refer to 
+   :ref:`<kotlin-model-data-device-sync>`. 
 
-You can check if the ``RealmSet`` contains one or more elements:
+Query By Property
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/crud/read.txt:275]: I agree and moved the Primary Key back and forth, but for some reason it seemed like it's elevated as an important query property in our docs and is harder to find in the OTP if it's under the Property section... 
But since you're also noting it, I'll go with my gut and move it back 😄

[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/crud/read.txt:275]: And re. the first half of your comment -- thank you! I kept getting stuck organizing this page as a single page with decent headers AND consistent terminology. But I think your suggestion fixes several of those things at once.

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 14]
```diff
@@ -1,201 +1,654 @@
...
+  `RealmResults <{+kotlin-local-prefix+}io.realm.kotlin.query/-realm-results/index.html>`__ 
+  collection, which represents all database objects that match the query 
+  conditions. In general, you can can work with a results collection like any 
+  other `Kotlin Collection <https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-collection/>`__.
+
+- Use `asFlow() <{+kotlin-local-prefix+}io.realm.kotlin.query/-realm-element-query/as-flow.html>`__ 
+  to perform an **asynchronous query**. The SDK lazily subscribes to a 
+  `Kotlin Coroutine Flow <https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/>`__ 
+  that you can collect and iterate over or listen to for changes. You *cannot*
+  call ``asFlow()`` on a ``MutableRealm.query``.
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/crud/read.txt:121]: the parenthetical latter is my understanding, but I can confirm

File: source/sdk/kotlin/realm-database/crud/read.txt (hunk 9 above)
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/crud/read.txt:556]: I'm definitely tweaking this based on how you did it for Swift

File: source/sdk/kotlin/realm-database/crud/read.txt (hunk 10 above)
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/crud/read.txt:560]: I considered grabbing it, but looking now, and seems like she's been able to get started on it

File: source/sdk/kotlin/realm-database/crud/read.txt (hunk 2 above)
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/crud/read.txt:164]: This was from the original page, but I can tweak the wording

File: examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/ReadTest.kt [hunk 15]
```diff
@@ -2,187 +2,892 @@ package com.mongodb.realm.realmkmmapp
...
+                }
+            }
+            // :snippet-end:
+            assertTrue(frozenFrogs.first().isFrozen())
+            val thrownException = assertFailsWith<Exception> {
+            // :snippet-start: frozen-vs-live-results
+            // 'Realm.query()' always returns frozen results
+            val frozenResults = realm.query<ExampleRealmObject_Frog>("age > $0", 50).find()
+            // If you try to modify the queried object, SDK throws 'IllegalStateException'
+            frozenResults.first().age += 1
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/ReadTest.kt:111]: I don't think we need to revisit it. This was just me brainlessly writing examples 😅

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 16]
```diff
@@ -1,201 +1,717 @@
...
+embedded object's ``coordinates`` property. A geospatial query checks 
+whether the point defined by the ``coordinates`` property is contained within
+the boundary of a defined geospatial shape.
+
+The SDK supports the following geospatial shapes:
+
+- ``GeoCircle``: defined by a center ``GeoPoint`` and a radius ``Distance``
+- ``GeoBox``: defined by two ``GeoPoint`` coordinates that represent the 
+  southwest and northeast corners of the box
+- ``GeoPolygon``: defined by a set of at least 4 ``GeoPoint`` coordinates that
```
Comment (dacharyc) [inline source/sdk/kotlin/realm-database/crud/read.txt:592]: Since we're not showing an example of querying with `GeoPolygon` here, should we mention that optionally this shape may contain holes that represent exclusive boundaries within the confines of the shape? Or do you think that's too weedy to get into here?

**Review by dacharyc — APPROVED** [review]
One small nit and a tiny Q, but this looks great! Nice work! 👏

File: source/sdk/kotlin/realm-database/crud/read.txt (hunk 16 above)
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/crud/read.txt:592]: probably worth mentioning without getting too in the weeds. I'll update

File: source/sdk/kotlin/realm-database/crud/read.txt [hunk 17]
```diff
@@ -1,201 +1,717 @@
...
-   :depth: 2
+   :depth: 3
    :class: singlecol
 
-.. tip:: ``find()`` is Synchronous
+This page describes how to query and lazily read objects persisted in a database 
+with the Atlas Device SDK for Kotlin. This lazy evaluation enables code 
+efficiency and performance when handling large data sets and complex queries. 
+
+.. note:: Cannot Read Asymmetric Objects
```
[the remark this answers is not in the document]
Comment (cbullinger) [reply source/sdk/kotlin/realm-database/crud/read.txt:26]: yep, will move 👍
